File: src/10-challenge/solution.js
/**
 * @param {string[]} instructions - The instructions to execute
 * @returns {number} The value of the register A
 */
export default function compile(instructions) {
	/** @type {{}} */
	const records = {}

	/** @type {Object.<'DEC' | 'INC' | 'MOV', (args: string) => void>} */
	const actions = {
		DEC: (args) => (records[args] = ~~records[args] - 1),
		INC: (args) => (records[args] = ~~records[args] + 1),

		MOV: (args) => {
			/** @type {RegExpMatchArray | null} */
			const match = args.match(/(-?\d+|\D) (\D)/)
			if (!match) return

			const [_, value, dest] = match
			records[dest] = value.match(/\D/) ? records[value] ?? 0 : parseInt(value)
		},
	}

	for (let i = 0; i < instructions.length; i++) {
		/** @type {string} */
		const instruction = instructions[i]

		/** @type {string} */
		const instructionAction = instruction.slice(0, 3)

		/** @type {string} */
		const instructionArgs = instruction.slice(4)

		if (instructionAction === 'JMP') {
			/** @type {RegExpMatchArray | null} */
			const match = instructionArgs.match(/(\D) (\d+)/)
			if (!match) continue

			const [_, record, newI] = match
			i = !records[record] ? parseInt(newI) - 1 : i
			continue
		}

		if (actions[instructionAction]) {
			actions[instructionAction](instructionArgs)
		}
	}

	return records['A']
}

File: src/10-challenge/solution.py
import re
from typing import Literal, Any

type Instruction = Literal["INC", "DEC", "JMP", "MOV"]


def compile(instructions: list[str]) -> int | None:
    records: dict[str, int] = {}

    for i, instruction in enumerate(iterable=instructions):
        instruction_action: str = instruction[:3]
        instruction_args: str = instruction[4:]

        if instruction_action == "JMP":
            match: re.Match[str] | None = re.search(
                pattern=r"(\D) (\d+)", string=instruction_args
            )
            if match is None:
                continue

            record: str | Any = match.group(1)
            new_i: str | Any = match.group(2)
            i: int = int(new_i) - 1 if records.get(record, 0) == 0 else i
            continue

        if instruction_action == "DEC":
            records.update(
                {f"{instruction_args}": records.get(instruction_args, 0) - 1}
            )
        elif instruction_action == "INC":
            records.update(
                {f"{instruction_args}": records.get(instruction_args, 0) + 1}
            )
        elif instruction_action == "MOV":
            match = re.search(pattern=r"(-?\d+|\D) (\D)", string=instruction_args)
            if match is None:
                continue

            value: str | Any = match.group(1)
            dest: str | Any = match.group(2)

            if re.search(pattern=r"\D", string=value) is None:
                records.update({f"{dest}": int(value)})
            else:
                records.update({f"{dest}": records.get(value, 0)})

    return records.get("A")
